addons/sale/models/product_product.py

```diff
@@ -1,23 +1,26 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from odoo import api, fields, models
+from odoo import fields, models
 
 
 class ProductProduct(models.Model):
     _inherit = 'product.product'
 
-    @api.multi
+    sales_count = fields.Integer(compute='_sales_count', string='# Sales')
+
     def _sales_count(self):
-        r = {}
-        domain = [
-            ('state', 'in', ['sale', 'done']),
-            ('product_id', 'in', self.ids),
-        ]
-        for group in self.env['sale.report'].read_group(domain, ['product_id', 'product_uom_qty'], ['product_id']):
-            r[group['product_id'][0]] = group['product_uom_qty']
+        self._cr.execute("""
+        SELECT  l.product_id,
+        sum(l.product_uom_qty / u.factor * u2.factor) AS product_uom_qty                    
+        FROM sale_order_line l
+           JOIN sale_order s ON l.order_id = s.id     
+           LEFT JOIN product_product p ON l.product_id = p.id
+           LEFT JOIN product_template t ON p.product_tmpl_id = t.id
+           LEFT JOIN product_uom u ON u.id = l.product_uom
+           LEFT JOIN product_uom u2 ON u2.id = t.uom_id
+        WHERE s.state in ('done', 'sale') and product_id in ({product_ids})
+        GROUP BY l.product_id""".format(product_ids=", ".join(str(i) for i in self.ids)))
+        product_qty = {r[0]: r[1] for r in self._cr.fetchall()}
```

Comment:
Not really faster, tried it and it was 90ms slower. Thx for hint.

Reply:
There is no way it is 90ms slower.  Simply impossible.  You cannot be measuring correctly. Are you using timeit?

Reply:
No, i measured it with the developer tools integrated in firefox. I think it doesn't metter, because it will only iterate the number of the products which will be read. in my example (see comment below) it was computing for 12 variants. But if you are sure it is much faster for more variants, i will modify and commit that.

Reply:
Well I just ran a test because I was like wtf.

On a list of tuples setup like this over a million iterations
```
list_of_tuples = [(x, randint(1,10000)) for x in range(1,100)]
timeit.timeit('dict(list_of_tuples)',globals={'list_of_tuples': list_of_tuples})
3.682828570003039
timeit.timeit('{r[0]: r[1] for r in list_of_tuples}',globals={'list_of_tuples': list_of_tuples})
6.105675358994631
```
So that is nanoseconds for each, hence why 90ms is impossible.  That said, I was surprised by how much faster dict comprehensions are in python 3.  In 2.7 they were dog slow.  Probably not worth worrying about, as the cost of the dict constructor on smaller lists will probably outweigh the benefit.

Reply:
>>> print timeit.timeit(stmt='''product_qty = {r[0]: r[1] for r in [(3062, 2616.0), (3066, 715.0), (3069, 406.0), (3071, 440.0), (3068, 174.0), (3064, 1126.0), (3058, 1729.0), (3067, 875.0), (3059, 47.0), (3070, 533.0), (3072, 293.0), (3061, 1.0), (3065, 382.0), (3063, 526.0)]} ''', number=10000)
0.0107140541077
>>> print timeit.timeit(stmt='''product_qty = dict([(3062, 2616.0), (3066, 715.0), (3069, 406.0), (3071, 440.0), (3068, 174.0), (3064, 1126.0), (3058, 1729.0), (3067, 875.0), (3059, 47.0), (3070, 533.0), (3072, 293.0), (3061, 1.0), (3065, 382.0), (3063, 526.0)]) ''', number=10000)
0.0293538570404

Reply:
In this method, I think there are max 80 items a call will be computed.

Reply:
@Guenzn 
is strange, i get the opposite, but as i say doesn't matter. Seems it was a 2.7 issue.

```
Python 3.6.5 (default, Apr  1 2018, 05:46:30) 
[GCC 7.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import timeit
>>> timeit.timeit(stmt='''product_qty = dict([(3062, 2616.0), (3066, 715.0), (3069, 406.0), (3071, 440.0), (3068, 174.0), (3064, 1126.0), (3058, 1729.0), (3067, 875.0), (3059, 47.0), (3070, 533.0), (3072, 293.0), (3061, 1.0), (3065, 382.0), (3063, 526.0)]) ''', number=10000)
0.02056067599914968
>>> timeit.timeit(stmt='''product_qty = {r[0]: r[1] for r in [(3062, 2616.0), (3066, 715.0), (3069, 406.0), (3071, 440.0), (3068, 174.0), (3064, 1126.0), (3058, 1729.0), (3067, 875.0), (3059, 47.0), (3070, 533.0), (3072, 293.0), (3061, 1.0), (3065, 382.0), (3063, 526.0)]} ''', number=10000)
0.03878806400462054
```

Reply:
@Guenzn oops this is v10 so 2.7